### ISSUE TITLE
New axes.set_prop_cycle method cannot handle any generic iterable

### ISSUE THREAD

**Issue by u55** [matplotlib/matplotlib#5368]
Dear matplotlib developers,

The docstring for the new method `axes.set_prop_cycle`, says that it can take any iterable for the property value.  However, I found that it fails with a ValueError if given a numpy.ndarray.  Here is a minimal working example:

```python
from __future__ import print_function
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import cm

cmap = cm.Set1(np.linspace(0,1,9))
print(type(cmap),'\n')

fig, ax = plt.subplots()
ax.set_prop_cycle('color', cmap.tolist())    # works
ax.set_prop_cycle('color', cmap)    # raises ValueError
plt.show()
```
which produces this traceback:

```
<type 'numpy.ndarray'> 

Traceback (most recent call last):
  File "/tmp/untitled.py", line 12, in <module>
    ax.set_prop_cycle('color', cmap)    # raises ValueError
  File "/usr/lib/python2.7/site-packages/matplotlib/axes/_base.py", line 1112, in set_prop_cycle
    prop_cycle = cycler(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/matplotlib/rcsetup.py", line 720, in cycler
    vals = validator(vals)
  File "/usr/lib/python2.7/site-packages/matplotlib/rcsetup.py", line 90, in f
    raise ValueError(msg)
ValueError: 's' must be of type [ string | list | tuple ]
```

Tested with matplotlib 1.5.0, python 2.7 and python 3.5, running in Arch Linux.
This failure in matplotlib 1.5.0 also affects the method `axes.set_color_cycle`, which is supposed to be deprecated but raises a ValueError instead.  The following code worked in matplotlib 1.4.3 but fails in 1.5.0:

```python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import cm

cmap = cm.Set1(np.linspace(0,1,9))

fig, ax = plt.subplots()
ax.set_color_cycle(cmap)    # raises ValueError
plt.show()
```

I think that, for convenience and for backwards compatibility with prior versions of `axes.set_color_cycle`, the new method `axes.set_prop_cycle` should be able to take *any* iterable, as the documentation suggests, and not be limited to only a string, list, or tuple.

Thank you,

**Comment by tacaswell** [comment]
Can you take a look at how this validation is done? The intention is definitely to take any iterable.

Attn @weathergod

**Comment by u55** [comment]
Yes, the validation comes about because `axes/_base.py` has the [line](https://github.com/matplotlib/matplotlib/blob/38782e31579a10faca98b348767b0231f041a7e9/lib/matplotlib/axes/_base.py#L36):
```python
from matplotlib.rcsetup import cycler
```
instead of using something such as:
```python
from cycler import cycler
```
and the version of `cycler` in `rcsetup.py` does the validating, including [this test](https://github.com/matplotlib/matplotlib/blob/38782e31579a10faca98b348767b0231f041a7e9/lib/matplotlib/rcsetup.py#L81):
```python
elif type(s) in (list, tuple):
```

Doing this sort of validating for the input parameters of `axes.set_prop_cycle` is unpythonic, in my opinion, because it goes against the principle of duck typing.

However, fixing the problem is not a simple matter of replacing `from matplotlib.rcsetup import cycler` with `from cycler import cycler`, because `axes.set_prop_cycle` has three different input [forms](https://github.com/matplotlib/matplotlib/blob/38782e31579a10faca98b348767b0231f041a7e9/lib/matplotlib/axes/_base.py#L1076-L1082):
```
Form 1 simply sets given `Cycler` object.
Form 2 creates and sets  a `Cycler` from a label and an iterable.
Form 3 composes and sets  a `Cycler` as an inner product of the
       pairs of keyword arguments. In other words, all of the
       iterables are cycled simultaneously, as if through zip().
```
and the [logic](https://github.com/matplotlib/matplotlib/blob/38782e31579a10faca98b348767b0231f041a7e9/lib/matplotlib/rcsetup.py#L688-L712) to handle this is currently in `rcsetup.cycler` rather than in `axes.set_prop_cycle`.

**Comment by tacaswell** [comment]
That test should probably be replaced with a try/except block once the
strings are excluded.

Can you take a crack at that?

**Comment by WeatherGod** [comment]
The fix is easy. Use cbook.isiterable instead of checking the types. That
check has always been there, but previously it was only applied to values
going through the rcparams route. When I rejiggered everything, I purposely
wanted validation everywhere and just used the available mechanism. I
didn't think about users supplying a ndarray of rgb values because I
focused on the rcparams codepath.

**Comment by u55** [comment]
@tacaswell @WeatherGod 
Question, how should `rcsetup._listify_validator` behave when passed an empty dictionary?

I tried replacing [this line](https://github.com/matplotlib/matplotlib/blob/38782e31579a10faca98b348767b0231f041a7e9/lib/matplotlib/rcsetup.py#L81):
```python
elif type(s) in (list, tuple):
    ...
```
with this line:
```python
import matplotlib.cbook as cbook
elif cbook.iterable(s):
    ...
```
which fixes the bug, but now [this test fails](https://github.com/matplotlib/matplotlib/blob/38782e31579a10faca98b348767b0231f041a7e9/lib/matplotlib/tests/test_rcparams.py#L284):
```python
...
        {'validator': validate_stringlist,
         'success': (('', []),
                     ('a,b', ['a', 'b']),
                     ('aardvark', ['aardvark']),
                     ('aardvark, ', ['aardvark']),
                     ('aardvark, ,', ['aardvark']),
                     (['a', 'b'], ['a', 'b']),
                     (('a', 'b'), ['a', 'b']),
                     ((1, 2), ['1', '2'])),
            'fail': ((dict(), ValueError),  # <------- this test fails
                     (1, ValueError),)
            },
...
```
Previously, an empty list would have been validated but an empty dictionary would have raised a ValueError.  An empty list must be valid since certain rcParams, [such as](https://github.com/matplotlib/matplotlib/blob/38782e31579a10faca98b348767b0231f041a7e9/lib/matplotlib/rcsetup.py#L1165) `'animation.ffmpeg_args'`, default to an empty list.  I see no reason that an empty dictionary should raise a ValueError, so should I remove this failing test or should I add an extra test on [this line](https://github.com/matplotlib/matplotlib/blob/38782e31579a10faca98b348767b0231f041a7e9/lib/matplotlib/rcsetup.py#L81) for an empty dictionary?

**Comment by WeatherGod** [comment]
That test has been in there for a long time, And I think an unordered
iterable is a bit silly. However, that probably doesn't mean that it is
correct to error out on it. The problem is creating a test that would pass
reliably on it. So, perhaps the code should be a bit tighter and test for
ordered iterables? Perhaps ABC's would help?

**Comment by tacaswell** [comment]
The dict should fail because this is expecting a sequence of values, not a mapping.

**Comment by u55** [comment]
Then I'm not sure how to do the test.  A numpy.ndarray reports that it is not an instance of `collections.Sequence`:

```python
>>> import collections
>>> import numpy as np
>>> isinstance(np.arange(9), collections.Sequence)
False
```

**Comment by u55** [comment]
@tacaswell  @WeatherGod 
I see no reason to reject unordered iterables, such as sets, however I can see why dictionaries are not really appropriate.
How about replacing [the validation test](https://github.com/matplotlib/matplotlib/blob/cb3c6f361376f14b10dae7027ba1e30b52f858a2/lib/matplotlib/rcsetup.py#L81) with this?:
```python
try:
    import collections.abc as abc
except ImportError:
    # python 2
    import collections as abc
...

...
        elif isinstance(s, abc.Iterable) and not isinstance(s, abc.Mapping):
            ...
```
This will correctly pass a numpy.ndarray and a generic generator.

**Comment by WeatherGod** [comment]
I am ok with that. I still don't like the idea of allowing sets and bags,
but I see no reason to prevent them, either.

**Comment by tacaswell** [comment]
In the case of `color_cycle` an un-ordered iterable does not make sense as there being an order is important to the function.

I think it is a problem that from process-to-process these validators will give different results.

**Comment by WeatherGod** [comment]
@tacaswell, note that the validator will convert whatever iterable into a
list, so it becomes ordered for the purpose of the cycler. Yes, which order
that will be is arbitrary, but that is for the user to decide if that is an
issue or not for their code.

**Comment by tacaswell** [comment]
I am just forseeing lots of very confused bug reports about randomized color cycles.

**Comment by WeatherGod** [comment]
Do you really think there are a lot of people keeping configuration values
around as sets?

Besides, I think this is a good compromise given that numpy's NDArray is
not a Sequence, so any check for orderability then falls to whitelisting
object types, which we know is doomed as soon as someone tries to pass a
pandas object through.

**Comment by tacaswell** [comment]
Do not underestimate users :stuck_out_tongue_winking_eye: 

It is probably worth letting that go through to keep the code simple, can we at least add a big developer warning in code comments summarizing this conversation?

**Comment by WeatherGod** [comment]
Sure:

```
# @WeatherGod: meh, let the users mess up themselves
# @tacaswell: SETS BAD!
```
;-)

**Comment by WeatherGod** [comment]
@u55, I hope the banter between @tacaswell and I didn't discourage you from putting together a pull request with your suggested fix.

**Comment by u55** [comment]
Yeah, no problem.  I'm just busy during the week. I'll submit a pull request soon.

**Comment by WeatherGod** [comment]
Fixed by #5452 and #5459
